src/pip/_internal/build_env.py

```diff
@@ -101,8 +106,55 @@ class SubprocessBuildEnvironmentInstaller:
     Install build dependencies by calling pip in a subprocess.
     """
 
-    def __init__(self, finder: PackageFinder) -> None:
+    def __init__(
+        self,
+        finder: PackageFinder,
+        build_constraints: list[str] | None = None,
+        build_constraint_feature_enabled: bool = False,
+        constraints: list[str] | None = None,
+    ) -> None:
         self.finder = finder
+        self._build_constraints = build_constraints or []
+        self._build_constraint_feature_enabled = build_constraint_feature_enabled
+        self._constraints = constraints or []
+
+    def _deprecation_constraint_check(self) -> None:
+        """
+        Check for deprecation warning: PIP_CONSTRAINT affecting build environments.
+
+        This warns when build-constraint feature is NOT enabled but regular constraints
+        match what PIP_CONSTRAINT environment variable points to.
+        """
+        if self._build_constraint_feature_enabled:
+            return
+
+        if self._build_constraints:
+            return
+
+        if not self._constraints:
+            return
+
+        if not os.environ.get("PIP_CONSTRAINT"):
+            return
+
+        pip_constraint_files = [
+            f for f in os.environ["PIP_CONSTRAINT"].split() if f.strip()
+        ]
+        if pip_constraint_files and pip_constraint_files == self._constraints:
+            deprecated(
+                reason=(
+                    "Setting PIP_CONSTRAINT will not affect "
+                    "build constraints in the future,"
+                ),
+                replacement=(
+                    "to specify build constraints use --build-constraint or "
+                    "PIP_BUILD_CONSTRAINT, to disable this warning without "
+                    "any build constraints set --use-feature=build-constraint or "
+                    'PIP_USE_FEATURE="build-constraint"'
```

Comment:
This reads poorly in the context of the full deprecation warning.

```
 DEPRECATION: Setting PIP_CONSTRAINT will not affect build constraints in the future, pip 26.2 will enforce this behaviour change. A possible replacement is to specify build constraints use --build-constraint or PIP_BUILD_CONSTRAINT, to disable this warning without any build constraints set --use-feature=build-constraint or PIP_USE_FEATURE="build-constraint"
```

```suggestion
                    "to specify build constraints using --build-constraint or "
                    "PIP_BUILD_CONSTRAINT. To disable this warning without "
                    "any build constraints set --use-feature=build-constraint or "
                    'PIP_USE_FEATURE="build-constraint"'
```

Perhaps?